### ISSUE TITLE
Kibana/ES logs instance breaks query tokens on punctuation

### ISSUE THREAD

**Issue by hannes-ucsc** [HumanCellAtlas/logs#45]
_From @hannes-ucsc on March 27, 2018 18:47_

Seaching by `cabe7031-3b80-4104-accb-23c08e561a51` results in matches that contain any of `cabe7031`, `3b80`, `4104`, `accb` or `23c08e561a51`. I would expect only matches containing exactly `cabe7031-3b80-4104-accb-23c08e561a51`. In a previous project that used SOLR (also based in Lucene) we had to modify the tokenizer and the query processor to break on whitespace only, not punctuation.

_Copied from original issue: HumanCellAtlas/data-store#1097_

**Comment by hannes-ucsc** [comment]
_From @Bento007 on March 28, 2018 22:54_

Use either the [pattern tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/5.5/analysis-pattern-tokenizer.html) or the [keyword tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/5.5/analysis-keyword-tokenizer.html).

**Comment by hannes-ucsc** [comment]
_From @parthshahva on March 28, 2018 23:36_

@Bento007 I'm actually planning on using the whitespace tokenizer (https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-whitespace-tokenizer.html). This will become the default tokenizer in an index template (used for all future indices created after this template is set up. Unfortunately is not retroactive to existing indices). cc @mweiden

**Comment by hannes-ucsc** [comment]
_From @parthshahva on March 28, 2018 23:45_

I'm also adding a default filter of lowercase. It should apply the whitespace tokenizer and lowercase filter both at indexing and at search time. Going forward, we can define specific tokenizers or filters at the mapping level, but I do not see the need for that at this time.

**Comment by hannes-ucsc** [comment]
_From @parthshahva on March 29, 2018 1:44_

This is now resolved in the dev instance, if you search by a quoted UUID, getting rid of one of the middle portions in the dash-separated-string will no longer match any log events. @hannes-ucsc and @Bento007 please play around with this for user acceptance.

Solution: Manually created an index template that fixes the dash issue by using whitespace analyzer in a dynamic index template that matches on all string fields

Current Status: Trying to figure out best way to automate creation of this index template in production as not to not require any manual steps.

**Comment by hannes-ucsc** [comment]
https://logs.data.youknowwhere349823984729834.org/_plugin/kibana/app/kibana#/discover?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:now%2Fd,mode:quick,to:now%2Fd))&_a=(columns:!(_source),index:'*',interval:auto,query:(query_string:(query:bc514c99-2a98-42c2-8617-b46b083dfa8d)),sort:!('@timestamp',desc))

The first result is a partial match for `8617`:

<img width="1071" alt="screen shot 2018-03-28 at 6 52 31 pm" src="https://user-images.githubusercontent.com/5143256/38065668-4bef65a2-32b9-11e8-8569-e98d35c4088b.png">

**Comment by hannes-ucsc** [comment]
_From @parthshahva on March 29, 2018 1:56_

@hannes-ucsc You have to wrap it in double quotes. I don't see a problem with allowing for dash separated searches unless someone wraps the string in quotes for an exact search. Is this different from your expectation?

**Comment by hannes-ucsc** [comment]
_From @parthshahva on March 29, 2018 1:59_

Lol, I don't think my template actually did anything... I didn't even touch prod and double quoting works there as well. I'll want to discuss with you why this is not enough/ an issue.

**Comment by hannes-ucsc** [comment]
Ah yes, with quotes it works. 

I'd prefer without quotes but if it's excessively complicated to get that working, with quotes would be acceptable. On Google, for example, the quotes are needed, too.

I seem (!) to remember that I tried UUIDs wrapped in quotes weeks ago and it didn't work. So this is progress!

**Comment by hannes-ucsc** [comment]
A logs search for a bundle UUID doesn't match log records that list the FQID (uuid.version). For example, this search by FQID matches

https://logs.dev.data.humancellatlas.org/_plugin/kibana/app/kibana#/discover?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:now-24h,mode:quick,to:now))&_a=(columns:!(_source),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'*',key:'@log_group',negate:!f,type:phrase,value:%2Faws%2Flambda%2Fdss-index-integration),query:(match:('@log_group':(query:%2Faws%2Flambda%2Fdss-index-integration,type:phrase))))),index:'*',interval:auto,query:(query_string:(query:%223bf64176-d7fd-494a-be85-e359bb7c2a43.2018-04-05T110612.631420Z%22)),sort:!('@timestamp',desc))

The corresponding search by UUID doesn't:

https://logs.dev.data.humancellatlas.org/_plugin/kibana/app/kibana#/discover?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:now-24h,mode:quick,to:now))&_a=(columns:!(_source),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'*',key:'@log_group',negate:!f,type:phrase,value:%2Faws%2Flambda%2Fdss-index-integration),query:(match:('@log_group':(query:%2Faws%2Flambda%2Fdss-index-integration,type:phrase))))),index:'*',interval:auto,query:(query_string:(query:%223bf64176-d7fd-494a-be85-e359bb7c2a43%22)),sort:!('@timestamp',desc))

I think it might be an issue. Can we fine-tune the tokenizer to break on `.` and whitespace?

**Comment by hannes-ucsc** [comment]
Or break on whitespace and any punctuation except dash?

**Comment by hannes-ucsc** [comment]
I just ran into this again and I completely forgot that I had already reported this. @mweiden and I just chatted about this and we agree that we'd should break on dot.

To summarize, a log message containing `ab12a3b2-365b-471d-91a4-c2f4fd796f96.2018-05-09T194558.930231Z` currently …

* … matches the query `"ab12a3b2-365b-471d-91a4-c2f4fd796f96.2018-05-09T194558.930231Z"` as it should
* ... doesn't match the query `"ab12a3b2-365b-471d-91a4-c2f4fd796f96"` but it should
* … matches the query `"ab12a3b2-365b-471d-91a4-c2f4fd796f96.2018"` which proves that dash and dot are treated differently.
